### PR TITLE
Adds the `rocm-mpich-base` container.

### PR DESCRIPTION
**Pull request by dipietrantonio**
I will prepare the equivalent docker files after having done more experimenting with it.

### CODE REVIEW

File: mpi/rocm-mpich-base/ubuntu22-rocm.def [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        apt-add-repository -y "deb https://apt.kitware.com/ubuntu/ jammy-rc main"
+        apt -y update
+        apt -y install cmake
+
+	[ -e /tmp/build ] && rm -rf /tmp/build	
+	mkdir -p /tmp/build
+	cd /tmp/build
+	
+	# Libfabric
+	wget https://github.com/ofiwg/libfabric/archive/refs/tags/v1.18.1.tar.gz
```
Comment (marcodelapierre) [inline mpi/rocm-mpich-base/ubuntu22-rocm.def:23]: I would parametrise the Libfabric version

File: mpi/rocm-mpich-base/ubuntu22-rocm.def [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      	&& ldconfig \
+      	&& cp -p /tmp/mpich-build/mpich-${MPICH_VERSION}/examples/cpi /usr/bin/ \
+      	&& cd / \
+      	&& rm -rf /tmp/mpich-build
+
+	cd /tmp/build
+        wget https://bootstrap.pypa.io/get-pip.py
+        python3 get-pip.py
+
+	wget https://repo.radeon.com/amdgpu-install/5.6/ubuntu/jammy/amdgpu-install_5.6.50600-1_all.deb
```
Comment (marcodelapierre) [inline mpi/rocm-mpich-base/ubuntu22-rocm.def:52]: I would parametrise the version of this AMD component (driver?)

File: mpi/rocm-mpich-base/ubuntu22-rocm.def [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	apt -y install ./amdgpu-install_5.6.50600-1_all.deb	
+	amdgpu-install -y --usecase=hiplibsdk,rocm,hip,opencl
+
+	# AWS OFI RCCL
+	cd /tmp/build
+        git clone https://github.com/ROCmSoftwarePlatform/aws-ofi-rccl.git
```
Comment (marcodelapierre) [inline mpi/rocm-mpich-base/ubuntu22-rocm.def:58]: any way of version controlling this dependency?

File: tensorflow/rocm-tensorflow.def [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	export HOROVOD_WITHOUT_PYTORCH=1
+	export HOROVOD_GPU=ROCM
+	export HOROVOD_GPU_OPERATIONS=NCCL
+	export HOROVOD_WITHOUT_GLOO=1
+	export HOROVOD_WITH_TENSORFLOW=1
+	export HOROVOD_ROCM_PATH=/opt/rocm
+	export HOROVOD_RCCL_HOME=/opt/rocm/rccl
+	export RCCL_INCLUDE_DIRS=/opt/rocm/rccl/include
+	export HOROVOD_RCCL_LIB=/opt/rocm/rccl/lib
+	export HCC_AMDGPU_TARGET=gfx90a
```
Comment (marcodelapierre) [inline tensorflow/rocm-tensorflow.def:17]: shall we parametrise the GPU generation?

File: tensorflow/rocm-tensorflow.def [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	python3 -m pip install --force-reinstall --upgrade --no-cache-dir "horovod==0.28.1"
```
Comment (marcodelapierre) [inline tensorflow/rocm-tensorflow.def:18]: I would parametrise the horovod version

File: mpi/rocm-mpich-base/ubuntu22-rocm.def [hunk 6]
```diff
@@ -0,0 +1,66 @@
+Bootstrap: docker
+From: ubuntu:22.04
+%post
+	ln -s /usr/share/zoneinfo/Australia/Perth /etc/localtime
+	apt -y update
+	apt -y upgrade
```
[the remark this answers is not in the document]
Comment (marcodelapierre) [reply mpi/rocm-mpich-base/ubuntu22-rocm.def:6]: this is also consistent with other recipes in this repo - the design choice was indeed made on the rationale explained above

**Review by marcodelapierre — DISMISSED** [review]
Great work!

General comment: I strongly recommend considering the migration to Dockerfiles ASAP, they are the de facto standard. We could then promptly generate SIF images after the Docker build, for HPC-ready compact images.

Both images built successfully.

`import tensorflow` in the Tensorflow image produced the following:
```
2023-08-28 06:38:58.329453: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
```
Can this be addressed in a time-effective way? Otherwise good to defer to later.

See comments across the codes.

**Comment by AlexisEspinosaGayosso** [comment]
@dipietrantonio , would it be possible to separate this as three different pull requests? One for the rocm-mpich-base.dockerfile, one for the pytorch and a third for tensorflow? In that way, a fix for one will not block the others (if the others are ready).

For example, I think we are in perfect position of pulling the rocm one. Indeed, I think this is sort of "urgent". As many other images (ours, a users made) can be built FROM this one. By the way, in that one, maybe we should stick to rocm/5.6 due to driver compatibility.

In regards to the PyTorch one, I did not read carefully, but can that and also the Tensorflow ones start FROM the rocm-mpich one using dockerfiles?

Cheers.

**Comment by dipietrantonio** [comment]
Hi @AlexisEspinosaGayosso ,
> would it be possible to separate this as three different pull requests?  
Yes I will do that, makes more sense.

> For example, I think we are in perfect position of pulling the rocm one. 

I am still working on that one, testing a few remaining things. I hope I can push it today to our Pawsey repository.

I will update this PR to be just about the `rocm-mpich-base`. I think the appropriate name should be `ubuntu:22-rocm5.6.0-mpich3.4.3` as this is mainly a base Ubuntu image, containing a few needed libraries. We should have a discussion about naming conventions.

**Comment by dipietrantonio** [comment]
@pelahi @AlexisEspinosaGayosso this PR is ready for review.

**Review by pelahi — COMMENTED** [review]
Is there a reason why this container doesn't build from the mpich lustre container?

**Comment by dipietrantonio** [comment]
I install libfabric as a dependency of https://github.com/ROCmSoftwarePlatform/aws-ofi-rccl , which is needed for optimal performance of the ROCm Communication Collective Library (RCCL). So this is a AMD specific build. I might get away by installing lib fabric and aws-ofi-rccl on top of the mpich-lustre container. I will try.

**Review by pelahi — APPROVED** [review]

**Pull request merged**